### PR TITLE
Feat/sync cookie

### PR DESCRIPTION
**Pull request by Hossein-Mirazimi**
this MR about sync localStorage with cookie

### CODE REVIEW

**Comment by Atinux** [comment]
Why adding the cookie sync @Hossein-Mirazimi ?

**Comment by Hossein-Mirazimi** [comment]
> Why adding the cookie sync @Hossein-Mirazimi ?

because in some cases may need to get preference value from cookies instead of local storage
#186

**Comment by tasiotas** [comment]
Yep, for subdomain for example

**Comment by tasiotas** [comment]
As far as I understand GDPR, it does not specify technology used for storage at all. It never uses term "cookie" or "local storage", it's about storing data for given reason.

So even with current method of using local storage, GDPR does apply.

The exceptions are outlined [here](https://gdpr-info.eu/art-6-gdpr/), but bit hard to understand.

I would like to believe that color theme saved in cookie/storage is necessary to deliver what visitor has asked for. But is that accurate interpretation of EU law, I don't know.

**Comment by wokalek** [comment]
Really need this feature. But aware of GPRD must will be included when enable this setting.

**Comment by Hossein-Mirazimi** [comment]
please help me add the GPRD rule for this

I don't know how to add this rule in the color-mode module

because this repo is just for managing the theme

If we should policy to set cookies for the theme

developers must do it in their project

**Comment by Atinux** [comment]
It would be nice to have a GPRD expert to know what we should do about this

**Comment by hacknug** [comment]
According to GitHub it is not necessary: https://github.blog/2020-12-17-no-cookie-for-you/

**Comment by Hossein-Mirazimi** [comment]
please review my changes

File: src/runtime/plugin.client.ts [hunk 1]
```diff
@@ -71,6 +71,20 @@ export default defineNuxtPlugin((nuxtApp) => {
     })
   }
 
+  function setPreferenceToStorage (storageType: typeof storage, preference: string) {
+    switch (storageType) {
+      case 'cookie':
+        window.document.cookie = storageKey + '=' + preference
```
Comment (RebeccaStevens) [inline src/runtime/plugin.client.ts:77]: Need ability to pass extra cookie options such as `expires`, `max-age`, `path`, `samesite`, `secure`, etc.

**Comment by Jordan-Ellis** [comment]
Since this pull request adds the ability to choose the storage type, it would be great if you could also specify a custom storage provider. For example, I'm using ionic and would prefer to store this in the native settings instead of local storage.

Something along the lines of this. The only thing I'm not sure of is where you would place the provider.

```js
// Custom settings store. Could be pinia store, native settings library, etc.
const mySettings = useMySettings() 

// Settings provider passed to color mode
const colorModeSettingProvider = {
    getPreference() {
        return mySettings.get("theme")
    },
    setPreference(value) {
        mySettings.set("theme", value)
    }
}
```

**Comment by Atinux** [comment]
It is a nice idea @Jordan-Ellis but the biggest issue is that it is written inside the inject script: https://github.com/nuxt-modules/color-mode/blob/961a4ab96a1a0e63bee34a4b86041558a6f0519c/src/script.ts#L9

This code is added in the `<head>` of the project in order to properly handle static site generation to avoid having a color switch on hydration. This script can only be JavaScript code that run in the browser and don't have access to any of the Vue runtime.

**Comment by Jordan-Ellis** [comment]
Ah, I didn’t realize that! Should that be moved to a separate issue, or is that beyond the scope of the project right now?

**Comment by manniL** [comment]
> Ah, I didn’t realize that! Should that be moved to a separate issue, or is that beyond the scope of the project right now?

@Jordan-Ellis yup 👍

**Comment by manniL** [comment]
@Atinux @danielroe WDYT with regards to that PR? Could we merge it into main?

File: src/script.ts [hunk 2]
```diff
@@ -71,3 +71,26 @@
     return '<%= options.fallback %>'
   }
 })()
+
+// @ts-ignore
+function getStorageValue(storageType, storageKey) {
+  if (typeof window === 'undefined') return null;
```
Comment (manniL) [inline src/script.ts:77]:
```suggestion
  if (window?.document === 'undefined') return null;
```

The other code will break on build.

**Comment by manniL** [comment]
Another missing feature would be be adding cookie reading to `plugin.server.ts` to already have the info on the server (if a cookie is set).

**Comment by Nisthar** [comment]
I got this error when using subdomains in my nuxt app. The color scheme was working fine on the main domain but not on the subdomains. Are there any workarounds for this issue or should i stop using nuxtui atm?

**Comment by atinux** [comment]
Closing in favour of #221

**Pull request closed without merging**